### ISSUE TITLE
dice loss问题

### ISSUE THREAD

**Issue by Meryl-Fang** [PaddlePaddle/PaddleX#589]
问题类型：其它  
**问题描述**  

========================
我想知道 "dice loss is only applicable to one channel classification"是说只支持单通道图片吗？ 
以及想在PaddleX里用Lovasz Loss可以调用PaddleSeg里的loss吗

谢谢！

**Comment by FlyingQianMM** [comment]
"dice loss is only applicable to one channel classification"是指dice loss仅能用于两类分类。另外请问下，这个信息是使用哪个repo提示的呢？

目前PaddleX里面不能直接调用PaddleSeg的Lovasz loss，后续我们会将lovasz loss加上。

**Comment by Meryl-Fang** [comment]
> "dice loss is only applicable to one channel classification"是指dice loss仅能用于两类分类。另外请问下，这个信息是使用哪个repo提示的呢？
> 
> 目前PaddleX里面不能直接调用PaddleSeg的Lovasz loss，后续我们会将lovasz loss加上。

PaddleX提示的。
是这样，我现在在试图写一个多分类的dice loss, 

我以为只要把label换成one-hot-shot，循环读每个class的logit就可以计算，但一直不成功。

```
#自定义 multi-class dice loss
def multi_dice_loss(logit, label, num_classes= 11):
    
    label = fluid.layers.cast(label, 'int64')
    label_one_hot = fluid.layers.one_hot(input=label, depth=num_classes)
    
    dice_sum = 0
    for i in range(len(num_classes)):
        dice_sum += dice_loss(logit[:,i,:,:], label_one_hot[:,:,:,i],ignore_mask, epsilon=0.00001)
    return dice_sum
```

是我对logit的shape（NCHW）理解不对吗？

**Comment by Meryl-Fang** [comment]
> "dice loss is only applicable to one channel classification"是指dice loss仅能用于两类分类。另外请问下，这个信息是使用哪个repo提示的呢？
> 
> 目前PaddleX里面不能直接调用PaddleSeg的Lovasz loss，后续我们会将lovasz loss加上。

以及如何打印loss.py里比如softmax的输入：label和logit的shape和value呢？为什么print label.shape 都是[-1,1,-1,-1]； -1 是要推断出来的值，但怎么能有三个数都是-1呢？

label.numpy() 或者label.detach()报错都是
`
'numpy' should be called by imperative Varible in imperative mode, please use fluid.dygraph.guard() as context to run it in imperative mode
`

不好意思还在学习中，问题比较多，谢谢！

**Comment by FlyingQianMM** [comment]
支持多分类的dice loss目前PaddleSeg已经实现了，参考https://github.com/PaddlePaddle/PaddleSeg/blob/d7bfdfacb2f63a262cc0db40ba00712e9e22a4a0/paddleseg/models/losses/dice_loss.py#L34

目前PaddleX还是基于静态图实现的，如果在PaddleX这里使用多分类的dice loss，「paddle 2.0.0」下的实现如下：
```python
import paddle
import paddle.nn.functional as F
def dice_loss(logit, label, num_classes=2, ignore_mask=None, epsilon=0.00001):
    ignore_mask = paddle.cast(ignore_mask, 'float32')
    logit = logit * ignore_mask
    label = label * ignore_mask
    label = paddle.transpose(label, [0, 2, 3, 1])
    label_one_hot = F.one_hot(label, num_classes)
    label_one_hot = paddle.transpose(label_one_hot, [0, 3, 1, 2]) 
    label_one_hot = paddle.cast(label_one_hot, 'float32')
    logit = F.softmax(logit)    
    dims = (0,) + list(range(2, len(label.ndimension())))
    intersection = paddle.sum(logits * labels_one_hot, dims)
    cardinality = paddle.sum(logits + labels_one_hot, dims)
    dice_loss = (2. * intersection / (cardinality + epsilon)).mean()

    label.stop_gradient = True
    ignore_mask.stop_gradient = True

    return 1 - dice_loss
```
同时需要将此处强制设置为1的代码去掉：https://github.com/PaddlePaddle/PaddleX/blob/e62d40c1322eff35835b61629c1b932cbfe45686/paddlex/cv/nets/segmentation/deeplabv3p.py#L424-L425

静态图下打印tensor信息，可以使用paddle.static.Print，使用方法参考paddle API文档https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/fluid/layers/control_flow/Print_cn.html#print
